### DISCUSSION TITLE
Umzug von 1.6.18 auf Raspi zu 2.x auf Docker/Synology

### DISCUSSION THREAD

**Discussion by mivola** [camueller/SmartApplianceEnabler#480] [Installation und Betrieb]
Hallo,

ich habe aktuell erfolgreich den SAE 1.6.18 auf einem Raspi laufen. Nun möchte ich den Umzug zu Docker auf einem Synology NAS planen. Dazu habe ich schon [Docker_DE.md](https://github.com/camueller/SmartApplianceEnabler/blob/master/doc/Docker_DE.md) & https://github.com/camueller/SmartApplianceEnabler/discussions/446 konsultiert.

Meine Fragen sind aber erstmal etwas genereller:
- soll ich zuerst den SAE auf dem Raspi auf 2.x updaten? Oder kann ich die Appliances.xml & Device2EM.xml auch in der aktuellen Version für den SAE2.x zur Verfügung stellen?
- was passiert, wenn im Netzwerk 2 Instanzen des SAE (auf Raspi & Synology) laufen? Sollte man das vermeiden und immer nur eine Instanz laufen haben?
- ich habe bereits einen Mosquitto MQTT Broker im Docker Container auf dem Synology laufen; dessen IP würde ich dann in der WebUI des SAE unter /settings eintragen; oder kann man das auch in der Konfiguration oder JAVA_OPTS mitgeben?

Danke und VG
Micha

**Comment by camueller** [comment]
> soll ich zuerst den SAE auf dem Raspi auf 2.x updaten? Oder kann ich die Appliances.xml & Device2EM.xml auch in der aktuellen Version für den SAE2.x zur Verfügung stellen?

Du solltest die XML-Dateien von Deinem Raspberry Pi in das Docker-Volume übernehmen. Der SAE aktualisiert sie automatisch.

> was passiert, wenn im Netzwerk 2 Instanzen des SAE (auf Raspi & Synology) laufen? Sollte man das vermeiden und immer nur eine Instanz laufen haben?

Für jedes verwaltete Gerät darf es immer nur eine Instanz geben. Ich selbst nutze beispielsweise parallel 2 Raspberry Pi produktiv, aber diese verwalten jeweils unterschiedliche Geräte.

> ich habe bereits einen Mosquitto MQTT Broker im Docker Container auf dem Synology laufen; dessen IP würde ich dann in der WebUI des SAE unter /settings eintragen; oder kann man das auch in der Konfiguration oder JAVA_OPTS mitgeben?

Du musst sie unter `/settings` eintragen - `JAVA_OPTS` geht nicht. Bei der IP-Adresse musst Du beachten, dass sie für Container-zu-Container-Kommunikation funktioniert! Das ist nicht die IP-Adresse, welche Du dem Mosquitto-Container als externe IP-Adresse zuordnest. Nur falls Du Dich wundern solltest, wenn der SAE nicht mit Mosquitto kommunizieren kann.

**Reply by mivola** [reply]
> > ich habe bereits einen Mosquitto MQTT Broker im Docker Container auf dem Synology laufen; dessen IP würde ich dann in der WebUI des SAE unter /settings eintragen; oder kann man das auch in der Konfiguration oder JAVA_OPTS mitgeben?
> 
> Du musst sie unter `/settings` eintragen - `JAVA_OPTS` geht nicht. Bei der IP-Adresse musst Du beachten, dass sie für Container-zu-Container-Kommunikation funktioniert! Das ist nicht die IP-Adresse, welche Du dem Mosquitto-Container als externe IP-Adresse zuordnest. Nur falls Du Dich wundern solltest, wenn der SAE nicht mit Mosquitto kommunizieren kann.

Also das generelle Setup scheint schonmal funktioniert zu haben. Die WebUI läuft und im Mosquitto sehe ich eine neue topic: `sae/Event/SempDevice2EM` mit: `{"time":"2023-02-23T11:42:56.506605","type":"MqttMessage"}`

Allerdings sehe ich im Log viele Exceptions: 
`11:42:59.146 [cling-7] ERROR o.f.cling.transport.spi.DatagramIO - Exception sending datagram to: /fe80:0:0:0:9209:d0ff:fe2b:b6c9%6: java.io.IOException: Address family not supported by protocol java.io.IOException: Address family not supported by protocol at java.base/java.net.PlainDatagramSocketImpl.send(Native Method) at java.base/java.net.DatagramSocket.send(DatagramSocket.java:695) at org.fourthline.cling.transport.impl.DatagramIOImpl.send(DatagramIOImpl.java:159) at org.fourthline.cling.transport.impl.DatagramIOImpl.send(DatagramIOImpl.java:150) at org.fourthline.cling.transport.RouterImpl.send(RouterImpl.java:309) at org.fourthline.cling.protocol.async.ReceivingSearch.sendSearchResponseAll(ReceivingSearch.java:176) at org.fourthline.cling.protocol.async.ReceivingSearch.sendResponses(ReceivingSearch.java:137) at org.fourthline.cling.protocol.async.ReceivingSearch.execute(ReceivingSearch.java:105) at org.fourthline.cling.protocol.ReceivingAsync.run(ReceivingAsync.java:71) at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) at java.base/java.lang.Thread.run(Thread.java:829)`

Muss ich mich darum kümmern bevor ich mit der Migration der XML-Dateien weiter mache?

Danke!
Micha

**Comment by camueller** [comment]
Die Meldung besagt, dass UPnP nicht funktioniert, d.h. der SHM wird nicht mit des SAE kommunizieren können.

**Reply by mivola** [reply]
> Die Meldung besagt, dass UPnP nicht funktioniert, d.h. der SHM wird nicht mit des SAE kommunizieren können.

Also ich weiß natürlich nicht wie & warum: aber die Kommunikation klappt trotz dieser Fehlermeldung: die Werte kommen im Sunny Portal an!
Nun ist die Frage: macht es Sinn dieser Fehlermeldung weiter auf den Grund zu gehen? @camueller kann ich dir diesbzgl irgendwie helfen?

Danke und VG
Micha

**Reply by mivola** [reply]
Also irgendwie/wann scheint die Kommunikation dann doch nicht mehr funktioniert zu haben (oder ich habe unwissentlich eine Änderung eingebaut). Die Werte der letzten Zeit fehlen im Sunny Portal. Ich konnte den Fehler im Log nun mittels der Java-Option `-Djava.net.preferIPv4Stack=true` beseitigen und die Werte tauchen wieder im SunnyPortal auf.

**Comment by camueller** [comment]
> Nun ist die Frage: macht es Sinn dieser Fehlermeldung weiter auf den Grund zu gehen? @camueller kann ich dir diesbzgl irgendwie helfen?

Ich kann softwareseitig nichts ändern. Der Fehler hängt mit Deinem Setup zusammen. Wenn es trotzdem funktioniert, gut für Dich :-)

**Reply by mivola** [reply]
> > Nun ist die Frage: macht es Sinn dieser Fehlermeldung weiter auf den Grund zu gehen? @camueller kann ich dir diesbzgl irgendwie helfen?
> 
> Ich kann softwareseitig nichts ändern. Der Fehler hängt mit Deinem Setup zusammen. Wenn es trotzdem funktioniert, gut für Dich :-)

OK, mal schauen ob ich da noch etwas nachforschen kann...

Nur noch eine letzte Frage: aktuell läuft im Docker die Version "2.1.0" des SAE. Wenn ich die Umgebungsvariable SAE_VERSION zB auf "2.2.1" ändere, wird dann automatisch diese neue Version benutzt? Muss man dabei etwas beachten? Oder läuft das alles automatisch?

Danke und VG
Micha

**Comment by camueller** [comment]
> Wenn ich die Umgebungsvariable SAE_VERSION zB auf "2.2.1" ändere, wird dann automatisch diese neue Version benutzt? Muss man dabei etwas beachten? Oder läuft das alles automatisch?

Nein. Siehe https://github.com/camueller/SmartApplianceEnabler/blob/master/doc/Update_DE.md#docker

Es sieht so aus, als ob die Grundlagen von Docker noch nicht ganz klar sind :-)

**Reply by mivola** [reply]
> > Wenn ich die Umgebungsvariable SAE_VERSION zB auf "2.2.1" ändere, wird dann automatisch diese neue Version benutzt? Muss man dabei etwas beachten? Oder läuft das alles automatisch?
> 
> Nein. Siehe https://github.com/camueller/SmartApplianceEnabler/blob/master/doc/Update_DE.md#docker
> 
> Es sieht so aus, als ob die Grundlagen von Docker noch nicht ganz klar sind :-)

Naja, tatsächlich ist es mein erstes Projekt mit Docker bei dem es etwas tiefer geht, aber der Link hilft mir leider nicht wirklich weiter. Ich nutze kein Docker Compose sondern betreibe ein SAE-Only-Docker-Image auf meiner Synology (mit einem separaten MQTT Broker). btw: der Link zur docker-compose.yaml führt zu einem 404...

Wenn ich mir die [docker-compose.yaml](https://github.com/camueller/SmartApplianceEnabler/blob/master/run/etc/docker/compose/docker-compose.yml) anschaue, sehe ich aber auch nicht, wie ich die Version des SAE speizifieren kann. Oder ist die Version fest im benutzten Docker Image "avanux/smartapplianceenabler" verdrahtet und man müsste auf eine neue Version des Image warten: https://hub.docker.com/r/avanux/smartapplianceenabler-amd64/tags ? Dann frage ich mich allerdings wozu diese Umgebungsvariable gesetzt/benutzt wird.

Danke
Micha

**Comment by camueller** [comment]
Der verlinkte Abschnitt zum Thema Update des SAE und hier konkret bei Nutzung des Docker-Images hat 3 Zeilen von denen Du leider nur die ersten zwei Zeilen gelesen hast. Es kommt aber auf die dritte Zeile an:

"Ansonsten muss lediglich das Docker-Image in der gewünschten Version verwendet werden."

**Reply by mivola** [reply]
> "Ansonsten muss lediglich das Docker-Image in der gewünschten Version verwendet werden."

OK, und das macht man über die Angabe eines entsprechenden Tags in Zeile 26: https://github.com/camueller/SmartApplianceEnabler/blob/df1e6bc648394ad52eaf42d86849bc3d489b03ea/run/etc/docker/compose/docker-compose.yml#L26 durch Hinzufügen von zb":2.1.0", oder? Und solange es im Docker Hub kein neueres Image gibt, kann man auch keine neuere Version des SAE nutzen (es sei denn, man baut sich sein eigenes Docker Image)? 

Ich möchte dich nicht nerven, ich möchte nur die technischen Zusammenhänge verstehen...

Danke
Micha

**Comment by camueller** [comment]
Was meinst Du mit "Hinzufügen"? Du kannst das Image nicht ändern!! Du musst das Image in der entsprechenden Version "pullen". Abgesehen davon hat die neueste Version auch immer das "latest"-Tag, d.h. ohne Angabe eine Version bekommt man auch immer die neueste Version.

Aber das sind Docker-Basics, die man wissen sollte, wenn man mit Docker-Images arbeitet. Es gibt unzählige Tutorials zu dem Thema und natürlich die offizielle Docker-Dokumentation.

**Reply by mivola** [reply]
> Was meinst Du mit "Hinzufügen"? Du kannst das Image nicht ändern!! Du musst das Image in der entsprechenden Version "pullen". 

Mit "Hinzufügen" meine ich: wenn ich (aus welchen Gründen auch immer) eine bestimmte SAE-Version (nicht latest) nutzen will/muss, dann geht das nur über die Angabe des Tags bei der Definition des Images in der docker-compose.yaml.
